h5py/h5t.pyx

```diff
@@ -1464,57 +1463,58 @@ cdef TypeCompoundID _c_compound(dtype dt, int logical, int aligned):
     # Compound datatypes
 
     cdef hid_t tid
-    cdef TypeID type_tmp
-    cdef dtype dt_tmp
-    cdef size_t offset
-    cdef size_t offset_step = 0
+    cdef TypeID member_type
+    cdef dtype member_dt
+    cdef size_t member_offset = 0
 
-    cdef tuple names = dt.names
-    cdef dict fields = {}
-    cdef list offsets
+    cdef dict offsets = {}
+    cdef list fields = []
 
     # The challenge with correctly converting a numpy/h5py dtype to a HDF5 type
     # which is composed of subtypes has three aspects we must consider
-    # 1. numpy/h5py dtypes do not always have the same size and HDF5, even when
+    # 1. numpy/h5py dtypes do not always have the same size as HDF5, even when
     #   equivalent (can result in overlapping elements if not careful)
     # 2. For correct round-tripping of aligned dtypes, we need to consider how
-    #   much padding we need
+    #   much padding we need by looking at the field offsets
     # 3. There is no requirement that the offsets be monotonically increasing
+    #  (so we start by sorting the names as a function of increasing offset)
     #
     # The code below tries to cover these aspects
 
+    # Get offsets for each compound member
     for name, field in dt.fields.items():
-        dt_tmp = field[0]
-        offset = field[1]
-        fields[offset] = {
-            "name": name.encode('utf8') if isinstance(name, unicode) else name,
-            "dtype": dtype(dt_tmp),
-            "size": py_create(dt_tmp, logical=logical).get_size(),
-        }
-
-    offsets = list(sorted(fields))
-    # Set initial size to itemsize or last offset plus itemsize, whichever is
-    # bigger
-    tid = H5Tcreate(H5T_COMPOUND,
-        max(dt.itemsize, offsets[-1] + fields[offsets[-1]]["size"])
-    )
-
-    for i, offset in enumerate(offsets):
-        dt_tmp = fields[offset]["dtype"]
-        type_tmp = py_create(dt_tmp, logical=logical, aligned=aligned)
-        if aligned and type_tmp.get_size() > dt_tmp.itemsize:
+        offsets[name] = field[1]
+
+    # Build list of names, offsets, and types, sorted by increasing offset
+    # (i.e. the position of the member in the struct)
+    for name in sorted(dt.names, key=offsets.__getitem__):
+        field = dt.fields[name]
+        name = name.encode('utf8') if isinstance(name, unicode) else name
+
+        # Get HDF5 data types and set the offset for each member
+        member_dt = field[0]
+        member_offset = max(member_offset, field[1])
+        member_type = py_create(member_dt, logical=logical, aligned=aligned)
+        if aligned:
+            if member_offset > field[1] or member_dt.itemsize != member_type.get_size():
```

Comment:
Don't we only care if the member is too big, not too small? (in reference to `member_dt.itemsize != member_type.get_size()`)

Reply:
Thanks for checking this out. I think that if the compound is aligned, the type sizes *should* be equivalent.  I could envision something weird happening if the numpy and h5py sizes aren't the same but are padded to the same alignment. In the case of the vlen members that this tries to fix, the vlen `void*` in numpy is the size of two `void*` in HDF5. If (for whatever reason) your numpy array is padded so that the next member after a variable-length member is an extra `sizeof(void*)`, then I think we need to check that both the offset isn't too big and that the member is the same size.

But you did make me realize that there's an inconsistency in the final member; if I'm doing
```py
        member_offset = max(member_offset, field[1])
```
then I should also be doing after the loop
```py
    member_offset = max(member_offset, dt.itemsize)
```

I'll push that change.

Reply:
I'd be concerned if the numpy and h5py sizes were different when aligned, as that would imply some underlying bug in either numpy or hdf5 (as they should both be using cpu specific types, or well defined generic types, both of which should have a single representation), I was more concerned about unusual padding, but that would also imply a bug in either numpy or hdf5, so ignore my previous comment.